src/Common/src/System/Net/Sockets/SocketAPMExtensions.cs

```diff
@@ -829,6 +829,24 @@ public static int EndSendTo(this Socket socket, IAsyncResult asyncResult)
         {
             return TaskToApm.End<int>(asyncResult);
         }
+
+        // Behavior adapter.
+        private static ArraySegment<byte> CreateArraySegment(byte[] buffer, int offset, int size)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+
+            try
+            {
+                return new ArraySegment<byte>(buffer, offset, size);
+            }
+            catch (ArgumentException ae)
+            {
+                throw new ArgumentOutOfRangeException(ae.Message, ae);
```

Comment:
Can you help me to understand why we're doing this translation?  Is it that we used to throw ArgumentOutOfRange for this case, but the code was changed to use ArraySegment and its validation, and that validation is throwing a different exception type than we used to?  Seems like unfortunate additional overhead (e.g. the non-inlineable CreateArraySegment method, the try/catch block, etc.) for this, but I guess these methods are considered legacy and there's already so much overhead this will be a drop in the bucket.